update bpf troubleshoot (#2294)
tigera/docs: calico/operations/ebpf/troubleshoot-ebpf.mdx

  BPF enabled, starting BPF endpoint manager and map manager.
  ```

> - In eBPF mode, forwaring external client access to services (typically NodePorts) from node to node is implemented using VXLAN encapsulation.

Reviewer: > forwaring

As the dog said, typo.

  only. Use `kubectl get pod -o wide -n calico-system` to find the name of a
  `calico-node` pod and use the name in the following commands instead of
  `<calico-node-name>`.

  To run the tool, use:

  ```bash
  kubectl exec -n calico-system <calico-node-name> -- calico-node -bpf <args>
  ```

  For example, to show the tool's help:

  ```bash
  kubectl exec -n calico-system <calico-node-name> -- calico-node -bpf help

    Available Commands:
      arp          Manipulates arp
      cleanup      Removes all calico-bpf programs and maps
      completion   Generate the autocompletion script for the specified shell
      connect-time Manipulates connect-time load balancing programs
      conntrack    Manipulates connection tracking
      counters     Show and reset counters
      help         Help about any command
      ifstate      Manipulates ifstate
      ipsets       Manipulates ipsets
      nat          Manipulates network address translation (nat)
      policy       Dump policy attached to interface
      profiling    Show and reset profiling data
      routes       Manipulates routes
      version      Prints the version and exits

> 

Reviewer 2: Two things:
1. The code block isn't closed. This is causing the build to fail. 
2. Did you remove the Flags section intentionally? 

@tomastigera

Author: > Did you remove the Flags section intentionally?

yes, that was intentional, no value really

  - In GCP, the "Allow forwarding" option must be enabled. As with AWS, traffic through a load balancer does not
    work correctly with DSR because the load balancer is not consulted on the return path from the backing node.

# The `calico-node -bpf` tool

> To inspect $[prodname]'s internal datastructures, you can use the `calico-node -bpf` tool.

Reviewer 2: ```suggestion
To inspect $[prodname]'s internal data structures, you can use the `calico-node -bpf` tool.
```

# The `calico-bpf` tool

> To inspect $[prodname]'s internal datastructures, you can use the `calico-bpf` tool.

Reviewer: Wouldn't make more sense to say something like `-bpf` argument with `calico-node`?
I feel like `calico-bpf` tool is a developer term, and we don't mention it anywhere else.

- Run the tool as follows:

  ```bash
  kubectl exec -n calico-system calico-node-abcdef -- calico-node -bpf <args>
  ```

  For example, to show the tool's help:

  ```bash
  kubectl exec -n calico-system calico-node-abcdef -- calico-node -bpf help

  Available Commands:
    arp          Manipulates arp
    cleanup      Removes all calico-bpf programs and maps
    completion   Generate the autocompletion script for the specified shell
    connect-time Manipulates connect-time load balancing programs
    conntrack    Manipulates connection tracking
    counters     Show and reset counters
    help         Help about any command
    ifstate      Manipulates ifstate
    ipsets       Manipulates ipsets
    nat          Manipulates network address translation (nat)
    policy       Dump policy attached to interface
    profiling    Show and reset profiling data
    routes       Manipulates routes
    version      Prints the version and exits

  (Since the tool is embedded in the main `calico-node` binary the `--help` option is not available, but running
  `calico-node -bpf help` does work.)

>   For example, to dump the BPF conntrack table use:

Reviewer 2: ```suggestion
  For example, to dump the BPF conntrack table, use:
```

  TCP 192.168.58.7:49178 -> 10.111.57.87:80 -> 192.168.105.136:80  Active ago 4.486606371s CLOSED              <--- example of connection to service with per-packet NAT
  TCP 10.128.1.194:41513 -> 10.128.1.192:179  Active ago 26.442759238s ESTABLISHED                             <--- example of connection without NAT or with connect-time NAT
  TCP 192.168.58.7:42818 -> 10.111.57.87:80 -> 192.168.105.136:80  Active ago 1m15.987585857s CLOSED
  TCP 10.128.1.192:58208 -> 10.109.136.88:7443 -> 192.168.58.5:7443  Active ago 4.935017508s ESTABLISHED
  UDP 162.142.125.240:30603 -> 10.128.1.192:18989  Active ago 1m0.816678617s
  UDP 127.0.0.1:48611 -> 127.0.0.53:53  Active ago 17.789851961s
  ```

  Note that traffic originating within the cluster uses connect-time load balancing. By default, the connect-time
  load balancing is only enabled for TCP traffic. When connect-time load
  balancing is used, the conntrack table will not show the NAT resolution as
  that happens when the application calls `connect()`.

## Check if $[prodname] is dropping packets

  If you suspect that $[prodname] is dropping packets, you can use the

>   `calico-node -bpf` tool to check the BPF counters. Since the eBPF dataplane is split

Reviewer 2: ```suggestion
  `calico-node -bpf` tool to check the BPF counters. Since the eBPF data plane is split
```

  into programs that are attached to interfaces, you must check the counters on

>   the relevant interface. You can either dump counters for all interfaces or use `--iface=<ifacename>` to dump

Reviewer 2: Consider `<interface-name>` as a clearer placeholder text.

  Usage:
    calico-bpf [command]
  Available Commands:
    arp          Manipulates arp
    cleanup      Removes all calico-bpf programs and maps
    completion   Generate the autocompletion script for the specified shell
    connect-time Manipulates connect-time load balancing programs
    conntrack    Manipulates connection tracking
    counters     Show and reset counters
    help         Help about any command
    ifstate      Manipulates ifstate
    ipsets       Manipulates ipsets
    nat          Manipulates network address translation (nat)
    policy       Dump policy attached to interface
    profiling    Show and reset profiling data
    routes       Manipulates routes
    version      Prints the version and exits

  (Since the tool is embedded in the main `calico-node` binary the `--help` option is not available, but running
  `calico-node -bpf help` does work.)

  For example, to dump the BPF conntrack table use:

  ```
  kubectl exec -n calico-system calico-node-abcdef -- calico-node -bpf conntrack dump
  ...
  ```

## Debug access to services

  Inspect the BPF NAT table to verify that the service is correctly programmed. To dump the BPF NAT table:

  ```

>   kubectl exec -n calico-system calico-node-abcdef -- calico-node -bpf nat dump

Reviewer: Suggestion, let's use the DS make it more copy pastable.
```bash
kubectl exec -n calico-system ds/calico-node -- calico-node -bpf nat dump
```

Reviewer 3: TIL that this was possible.  Since these are troubleshooting steps though, I think the user may want to be specific about which node they're looking at (e.g. where their application is).

Reviewer 2: I'd use placeholder values here. Something like:

`kubectl exec -n calico-system <calico-node-name> -- calico-node -bpf nat dump`

After the code block:

Replace `<calico-node-name>` with the name of the node you're debugging.

Author: Sure, I can make it that way. Just would make a note at the start of the doc, not after every example.

  10.109.136.88 port 7443 proto 6 id 1 count 1 local 0
	  1:0	 192.168.129.72:7443
  10.109.139.39 port 443 proto 6 id 7 count 2 local 0
	  7:0	 192.168.129.67:5443
	  7:1	 192.168.129.69:5443
  10.96.0.1 port 443 proto 6 id 3 count 1 local 0
	  3:0	 10.128.0.255:6443

  ```

  Inspect the BPF conntrack table to verify that connections are being tracked. To dump the BPF conntrack table:

  ```
  kubectl exec -n calico-system calico-node-abcdef -- calico-node -bpf conntrack dump

  TCP 192.168.58.7:49178 -> 10.111.57.87:80 -> 192.168.105.136:80  Active ago 4.486606371s CLOSED              <--- example of connection to service with per-packet NAT
  TCP 10.128.1.194:41513 -> 10.128.1.192:179  Active ago 26.442759238s ESTABLISHED                             <--- example of connection without NAT or with connect-time NAT
  TCP 192.168.58.7:42818 -> 10.111.57.87:80 -> 192.168.105.136:80  Active ago 1m15.987585857s CLOSED
  TCP 10.128.1.192:58208 -> 10.109.136.88:7443 -> 192.168.58.5:7443  Active ago 4.935017508s ESTABLISHED
  UDP 162.142.125.240:30603 -> 10.128.1.192:18989  Active ago 1m0.816678617s
  UDP 127.0.0.1:48611 -> 127.0.0.53:53  Active ago 17.789851961s
  ```

>   Note that traffic originating within the cluster uses connect-time load balancing. By default, the connect-time

Reviewer: This could be written with `>` at the start of each sentence out side the code block ideally at the start of this block.

  load balancing is only enabled for TCP traffic. When connect-time load
  balancing is used, the conntrack table will not show the NAT resolution as
  that happens when the application calls `connect()`.

## Check if $[prodname] is dropping packets

  If you suspect that $[prodname] is dropping packets, you can use the
  `calico-bpf` tool to check the BPF counters. Since the eBPF dataplane is split
  into programs that are attached to interfaces, you must check the counters on

>   the relevant interface. You can either dump counters for all interfaces or user `--iface=<ifacename>` to dump

Reviewer: >  for all interfaces or user

or use?

  - A host endpoint ingress program is executed on the path from external node to the host.
  - A host endpoint egress program is executed on the path from host to external host.

- you may also see `ens192---X` which indicates an XDP program. $[prodname] uses XDP programs to implement `doNotTrack` policies on host devices only.

- `Final result=ALLOW (-1). Program execution time: 7366ns` is the message. In this case, logging the final result of
  the program. Note that the timestamp is massively distorted by the time spent logging.

## Debugging policy issues

If you suspect that $[prodname] is dropping packets due to policy, you can use the
`calico-node -bpf` tool to dump the policy that is attached to a specific interface.

```
kubectl exec -n calico-system calico-node-abcdef -- calico-node -bpf policy dump <interface> <hook>

> where `<interface>` is the name of the interface and `<hook>` is either `ingress`, `egress`, `xdp` or `all`.

Reviewer 2: Move outside the code block, like this:

Replace the following:
* `<interface>`: The name of the interface.
* `<hook>`: description of what this is. Must be one of the following: `ingress`, `egress`, `xdp`, or `all`.

```

Dump of an ingress policy. Note that ingress policy for a pod is attached to the egress hook of the host-side of the caliX veth pair.

```
IfaceName: calic31b4f7fc58
Hook: tc egress
Error:
Policy Info:
// Start of tier default
// Start of policy default/knp.default.allow-nginx-from-ubuntu
// Start of rule action:"allow"  protocol:{name:"tcp"}  dst_ports:{first:80  last:80}  src_ip_set_ids:"s:nzE8vwTu69FSscx2FDKjb20D9dZxEyVxsWFqwA"  original_src_selector:"projectcalico.org/orchestrator == 'k8s' && app == 'ubuntu-client'"  rule_id:"29vMYcPWr7reSxxN"
// IPSets src_ip_set_ids:<0x303904ae5eae5418>
// count = 9
// End of rule 29vMYcPWr7reSxxN
// End of policy default/knp.default.allow-nginx-from-ubuntu
// End of tier default: deny
// Start of rule action:"allow"  rule_id:"aBMQCbsUMESPKGRp"
// count = 0
// End of rule aBMQCbsUMESPKGRp
```

> Rules that use selectors refer to IP sets. You can dump the contents of an IP set using the `ipsets` command and you can check whether the IP set contains the expected members:

Reviewer 2: I'm not sure about the colon. Was there meant to be a code example after this?

You can see how many packets have matched each rule. In this example, 9 packets have matched rule `29vMYcPWr7reSxxN` and 0 packets have matched rule `aBMQCbsUMESPKGRp`.

Adding `--asm` will show the eBPF assembly code for the program as well.

```
kubectl exec -n calico-system calico-node-abcdef -- calico-node -bpf ipsets dump

IP set 0x303904ae5eae5418
   192.168.58.5/32

IP set 0xffef9f925a8a4ca4
   192.168.129.66/32
   192.168.129.68/32
```

## Debugging calico-node not ready

If you notice that a `calico-node` pod is not ready, check its logs for errors. The most likely reason for `calico-node` not being ready in eBPF mode is
that $[prodname] is not able to update a program attached to an interface. Look for the following type of warning:

```

> 2025-09-22 22:39:59.801 [WARNING][10374] felix/bpf_ep_mgr.go 2107: Failed to apply policy to endpoint, leaving it dirty error=<error message>

Reviewer 2: Better to use an actual error message here, since you're looking at a specific warning.

Alternatively, you can cut off the sample earlier.
```
2025-09-22 22:39:59.801 [WARNING][10374] felix/bpf_ep_mgr.go 2107: Failed to apply policy to endpoint, leaving it dirty

```

> One reason for this type of error is that the eBPF programs provided in the $[nodecontainer] image are not compatible with the verifier used by your kernel. Each kernel must ensure that the eBPF programs it loads are safe to run, however,

Reviewer 2: ```suggestion
One reason for this type of error is that the eBPF programs provided in the $[nodecontainer] image are not compatible with the verifier used by your kernel. Each kernel must ensure that the eBPF programs it loads are safe to run. However,
```
easier to parse

> capabilities of the verifier vary between kernel versions. We do test the eBPF programs with a range of kernels, but it is impossible to test all kernels. You may see errors such as if the verifier rejects a program:

Reviewer 2: ```suggestion
capabilities of the verifier differ between kernel versions. We do test the eBPF programs with a range of kernels, but it is impossible to test all kernels. You may see errors such as if the verifier rejects a program:
```
Avoiding repetition

- If you are running your cluster in a cloud such as AWS, then your cloud provider may limit the bandwidth between
  nodes in your cluster. For example, most AWS nodes are limited to 5GBit per connection.

## Debug high CPU usage

If you notice `$[noderunning]` using high CPU:

- Check if `kube-proxy` is still running. If `kube-proxy` is still running, you must either disable `kube-proxy` or
  ensure that the Felix configuration setting `bpfKubeProxyIptablesCleanupEnabled` is set to `false`. If the setting
  is set to `true` (its default), then Felix will attempt to remove `kube-proxy`'s iptables rules. If `kube-proxy` is

>   still running, it will fight with `Felix`.

Reviewer: maybe expand on `it will fight with Felix` or remove it. I do love it tho 😆

- If you are running your cluster in a cloud such as AWS, then your cloud provider may limit the bandwidth between
  nodes in your cluster. For example, most AWS nodes are limited to 5GBit per connection.

### Runtime profiling

> Release 3.31 introduced runtime profiling of the eBPF programs. The profiling data can be examined using the

Reviewer 2: Remove mention of a specific release. Since this is the 3.31 documentation, you can just talk about the feature as a standard thing.
